python/ml4ir/base/stats/t_test.py

```diff
@@ -0,0 +1,312 @@
+    """
+    for i in range(len(diff)):
+        count += 1
+        delta = diff[i] - mean
+        mean += delta / count
+        delta2 = diff[i] - mean
+        m2 += delta * delta2
+    return count, mean, m2
+
+
+def t_test_log_results(t_test_stat, pvalue, ttest_pvalue_threshold, logger):
+    """
+    performing click rank distribution t-test
+
+    Parameters
+    ----------
+    t_test_stat: float
+        The t-test statistic
+    pvalue: float
+        The p-value of the t-test statistic
+    ttest_pvalue_threshold: float
+        The p-value threshold
+    logger: Logger
+        Logger object to log t-test significance decision
+
+    """
+    logger.info(
+        "Performing a paired t-test between the click rank distribution of new model and the old model:\n\tNull hypothesis: There is no difference between the two click distributions.\n\tAlternative hypothesis: There is a difference between the two click distributions")
+    logger.info("t-test statistic={}, p-value={}".format(t_test_stat, pvalue))
+    if pvalue < ttest_pvalue_threshold:
+       logger.info(
+            "With p-value threshold={} > p-value --> we reject the null hypothesis. The click rank distribution of the new model is significantly different from the old model".format(
+                ttest_pvalue_threshold))
+    else:
+       logger.warning(
+            "With p-value threshold={} < p-value --> we cannot reject the null hypothesis. The click rank distribution of the new model is not significantly different from the old model".format(
+                ttest_pvalue_threshold))
+
+
+def run_ttest(mean, variance, n, ttest_pvalue_threshold, logger):
+    """
+    Compute the paired t-test statistic and its p-value given mean, standard deviation and sample count
+    Parameters
+    ----------
+    logger: Logger
+        Logger object to log t-test significance decision
+    mean: float
+        The mean of the rank differences for the entire dataset
+    variance: float
+        The variance of the rank differences for the entire dataset
+    n: int
+        The number of samples in the entire dataset
+    ttest_pvalue_threshold: float
+        P-value threshold for student t-test
+
+    Returns
+    -------
+    t_test_metrics_dict: Dictionary
+        A dictionary with the t-test metrics recorded.
+    """
+    t_test_stat, pvalue = perform_click_rank_dist_paired_t_test(mean, variance, n)
+    t_test_log_results(t_test_stat, pvalue, ttest_pvalue_threshold, logger)
+
+    # Log t-test statistic and p-value
+    t_test_metrics_dict = {'Rank distribution t-test statistic': t_test_stat,
+                           'Rank distribution t-test pvalue': pvalue,
+                           'Rank distribution t-test difference is statistically significant': pvalue < ttest_pvalue_threshold}
+    return t_test_metrics_dict
+
+
+def statistical_analysis_preprocessing(clicked_records, group_metric_running_variance_params, group_key, variance_list):
```

Comment:
@mohazahran Rename the method to something that mentions computing mean and variance for a batch.

Reply:
ping